lib/Macros/Sources/SwiftMacros/DebugDescriptionMacro.swift

```diff
@@ -0,0 +1,312 @@
+                    let message: ErrorMessage = "cannot reference computed properties"
+                    context.diagnose(node: expr, message: message)
+                    return []
+                }
+
+                let propertyPath = propertyChain.joined(separator: ".")
+                summarySegments.append("${var.\(propertyPath)}")
+            @unknown default:
+                return []
+            }
+        }
+
+        let summaryString = summarySegments.joined()
+
+        var typeIdentifier: String
+        if let typeParameters = genericParameters(declaration), typeParameters.count > 0 {
+            let typePatterns = Array(repeating: ".+", count: typeParameters.count).joined(separator: ",")
+            // A regex matching that matches the generic type.
+            typeIdentifier = "^\(moduleName)\\.\(typeName)<\(typePatterns)>"
+        } else if declaration.is(ExtensionDeclSyntax.self) {
+            // When attached to an extension, the type may or may not be a generic type.
+            // This regular expression handles both cases.
+            typeIdentifier = "^\(moduleName)\\.\(typeName)(<.+>)?$"
+        } else {
+            typeIdentifier = "\(moduleName).\(typeName)"
+        }
+
+        // Serialize the type summary into a global record, in a custom section, for LLDB to load.
+        let decl: DeclSyntax = """
+        #if os(Linux)
+        @_section(".lldbsummaries")
+        #elseif os(Windows)
+        @_section(".lldbsummaries")
+        #else
+        @_section("__DATA_CONST,__lldbsummaries")
+        #endif
+        @_used
+        let \(raw: typeName)_lldb_summary = (
+            \(raw: encodeTypeSummaryRecord(typeIdentifier, summaryString))
+        )
+        """
+
+        return [decl]
+    }
+}
+
+// MARK: - Diagnostics
+
+private struct ErrorMessage: DiagnosticMessage, ExpressibleByStringInterpolation {
+    init(stringLiteral value: String) {
+        self.message = value
+    }
+    var message: String
+    var diagnosticID: MessageID { .init(domain: "DebugDescription", id: "DebugDescription")}
+    var severity: DiagnosticSeverity { .error }
+}
+
+extension MacroExpansionContext {
+    fileprivate func diagnose(node: some SyntaxProtocol, message: any DiagnosticMessage) {
+        diagnose(Diagnostic(node: node, message: message))
+    }
+}
+
+// MARK: - AST Helpers
+
+private func extractModuleName(from fileID: ExprSyntax) -> String? {
+    if let fileID = fileID.as(StringLiteralExprSyntax.self)?.representedLiteralValue,
+       let firstSlash = fileID.firstIndex(of: "/") {
+        return String(fileID.prefix(upTo: firstSlash))
+    }
+    return nil
+}
+
+private func identifier<Decl: DeclSyntaxProtocol>(of decl: Decl) -> TokenSyntax? {
+    decl.as(StructDeclSyntax.self)?.name ??
+    decl.as(ClassDeclSyntax.self)?.name ??
+    decl.as(EnumDeclSyntax.self)?.name ??
+    decl.as(ExtensionDeclSyntax.self)?.extendedType.as(IdentifierTypeSyntax.self)?.name
+}
```

Comment:
I hear you. I understand that extensions are common, but I made these free functions because I felt that functions provide a way to distinguish between the API of the type, and the non-API support functions I've written locally.

Using the example `decl.identifier`, to someone well versed in SwiftSyntax, they may know that `identifier` is not a core part of the API of `decl`. But I don't have SwiftSyntax's API memorized, and so if I come back to this code later, I can't tell just by looking at a line, what's API, and what's local extension. I would prefer not to write property chains that intermix API with local extensions. My thinking is that it helps me to understand the code better if I can see the boundary between API and non-API.

Obviously this is subjective, I am willing to change this, but I did this with the above thoughts.

Reply:
As far as I know (and the way I have been writing Swift code for the last years) defining these local extensions is the best practice. But it’s your code, so ultimately your decision.